vars/scmNotify.groovy

```diff
@@ -0,0 +1,31 @@
+// vars/scmNotify.groovy
+
+ /* This provides a way of notifying the SCM such as GitLab/GitHub with
+  * the status of the build stage in progress.
+  *
+  * The scmNotifySystem must be provided as a shared
+  * groovy library local to the running Jenkins for this routine.
+  *
+  * If it is not provided this routine will not do anything.
+  */
+
+  /**
+   * scmNotify step method
+   *
+   * @param config Map of parameters passed
+   *
+   * See the githubNotify pipeline step for the parameters to pass.
+   */
+
+def call(Map config = [:]) {
+
+  if (env.DAOS_JENKINS_NOTIFY_STATUS == null) {
+    println "Jenkins not configured to notify SCM repository of builds."
+    return
+  }
+  try {
+    scmNotifySystem(config)
```

Comment:
Why is this needed?  I'm not OK with removing even more functionality from the open code and burying it into closed modules, particularly when it results in the open code becoming less functional that it was before.

Reply:
This does three purposes per the README.md file.
1. It makes the Jenkinsfile just work on systems that use our repository but do not have permission to notify it with out needing a Jenkins environment variable to inhibit updates.
2. It isolates if building on systems that are not checked out from Github that can still get feedback, such as Gitlab if the site chooses to provide their own additional pipeline library.
3. It removes requiring the implementation that other sites use the same credential-id for the scm access and at the same time removes the need to expose the credential-id.

Reply:
But it also disables notifying a GitHub fork that somebody might make and want to duplicate our CI workflow.  In that regard, this is a functionality regression.

But also, if we have code to update either a GitHub fork or a fork in a Gitlab, why wouldn't we want the open project to benefit from that code instead of hiding it away?

Reply:
And as I have said before, _hiding_ credential-ids is security-through-obscurity (which means little to no security and should not be relied on as it only provides a false sense of security).  The security mechanism that should be applied to the use of credential-ids is pre-execution review gating of PRs, not obscurity.

Reply:
It only disables notifying a Github fork if the end user has not setup their own additional pipeline-lib.

Previous implementation restricted the SCM hosting to GitHub and also the required the end user to use the same Jenkins credential ID as we were using.  Or it required them setting an environment variable.

This patch removes those restrictions, making the project more open and portable.

Reply:
Not the forum for discussing security issues.

Reply:
Changed code to first try the scmNotifySystem library and if it is not present, use the environment variable that enables SCM reporting as the source of a credential-id that is used for a githubNoify call, to allow an alternate way for a site to be configured.